gobblin-example/src/main/java/gobblin/example/wikipedia/WikipediaSource.java

```diff
@@ -41,13 +50,53 @@
  */
 public class WikipediaSource extends AbstractSource<String, JsonElement> {
 
+  public static final String ARTICLE_TITLE="gobblin.wikipediaSource.workUnit.title";
+
   @Override
   public List<WorkUnit> getWorkunits(SourceState state) {
+
+    Map<String, Iterable<WorkUnitState>> previousWorkUnits = state.getPreviousWorkUnitStatesByDatasetUrns();
+    List<String> titles = new LinkedList<>(Splitter.on(",").omitEmptyStrings().
+        splitToList(state.getProp(WikipediaExtractor.SOURCE_PAGE_TITLES)));
+
+    Map<String, LongWatermark> prevHighWatermarks = Maps.newHashMap();
+    for (Map.Entry<String, Iterable<WorkUnitState>> entry : previousWorkUnits.entrySet()) {
+      Iterable<LongWatermark> watermarks =
+          Iterables.transform(entry.getValue(), new Function<WorkUnitState, LongWatermark>() {
+        @Override
+        public LongWatermark apply(WorkUnitState wus) {
+          return wus.getActualHighWatermark(LongWatermark.class);
+        }
+      });
+      watermarks = Iterables.filter(watermarks, Predicates.notNull());
+      List<LongWatermark> watermarkList = Lists.newArrayList(watermarks);
+      Collections.sort(watermarkList, Collections.<LongWatermark>reverseOrder());
```

Comment:
This looks like an expensive way to get the max. Why not use Collections.max() ?